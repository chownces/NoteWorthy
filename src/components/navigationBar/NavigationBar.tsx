import React from 'react';
import { useMediaQuery } from 'react-responsive';
import { Link } from 'react-router-dom';
import { Icon, Image, Menu, Sidebar } from 'semantic-ui-react';

import brandLogo from '../../assets/brand_logo.png';

enum NavbarItems {
  allNotes,
  login,
  todos,
  contribute,
  boardView
}

const NavigationBar: React.FC = () => {
  const [activeItem, setActiveItem] = React.useState<NavbarItems>(NavbarItems.allNotes);
  const [sidePanelOpen, setSidePanelOpen] = React.useState<boolean>(false);
  const isMobileBreakpoint = useMediaQuery({ maxWidth: 768 });

  const setActiveItemAndCloseSidePanel = (activeItem: NavbarItems) => {
    setActiveItem(activeItem);
    setSidePanelOpen(false);
  };

  // NOTE: We use Link instead of href inside Menu.item to prevent unnecessary page refresh
  const desktopMenuItems = (
    <>
      <Menu>
        <Link to="/" className="brand-logo-link">
          <Menu.Item
            as="div"
            active={false}
            onClick={() => setActiveItem(NavbarItems.allNotes)}
            className="brand-logo-container"
          >
            <Image src={brandLogo} size="small" />
          </Menu.Item>
        </Link>
        <Link to="/">
          <Menu.Item
            as="div"
            active={activeItem === NavbarItems.allNotes}
            onClick={() => setActiveItem(NavbarItems.allNotes)}
          >
            <Icon name="file alternate outline" />
            All Notes
          </Menu.Item>
        </Link>
        <Link to="/todos">
          <Menu.Item
            as="div"
            active={activeItem === NavbarItems.todos}
            onClick={() => {
              setActiveItem(NavbarItems.todos);
            }}
          >
            <Icon name="calendar check outline" />
            Todo List
          </Menu.Item>
        </Link>
        <Link to="/BoardView">
          <Menu.Item
            as="div"
            active={activeItem === NavbarItems.boardView}
            onClick={() => {
              setActiveItem(NavbarItems.boardView);
            }}
          >
            BoardView
          </Menu.Item>
        </Link>
        <Menu.Menu position="right">
          <Link to="/contribute">
            <Menu.Item
              as="div"
              active={activeItem === NavbarItems.contribute}
              onClick={() => setActiveItem(NavbarItems.contribute)}
            >
              Contribute
            </Menu.Item>
          </Link>
          <Link to="/login">
            <Menu.Item
              as="div"
              active={activeItem === NavbarItems.login}
              onClick={() => setActiveItem(NavbarItems.login)}
            >
              Login
            </Menu.Item>
          </Link>
        </Menu.Menu>
      </Menu>
    </>
  );

  const mobileTopMenuItems = (
    <Menu>
      <Menu.Item as="a" onClick={() => setSidePanelOpen(true)}>
        <Icon name="bars" />
      </Menu.Item>
    </Menu>
  );

  const mobileSideMenuItems = (
    <Sidebar
      as={Menu}
      animation="overlay"
      onHide={() => setSidePanelOpen(false)}
      visible={sidePanelOpen}
      vertical
    >
      <Link to="/">
        <Menu.Item
          as="div"
          active={activeItem === NavbarItems.allNotes}
          onClick={() => setActiveItemAndCloseSidePanel(NavbarItems.allNotes)}
        >
          <Icon name="file alternate outline" />
          All Notes
        </Menu.Item>
      </Link>
      <Link to="/todos">
        <Menu.Item
          as="div"
          active={activeItem === NavbarItems.todos}
          onClick={() => setActiveItemAndCloseSidePanel(NavbarItems.todos)}
        >
          <Icon name="calendar check outline" />
          Todo List
        </Menu.Item>
      </Link>
      <Link to="/contribute">
        <Menu.Item
          as="div"
          active={activeItem === NavbarItems.contribute}
          onClick={() => setActiveItemAndCloseSidePanel(NavbarItems.contribute)}
        >
          Contribute
        </Menu.Item>
      </Link>
      <Link to="/login">
        <Menu.Item
          as="div"
          active={activeItem === NavbarItems.login}
          onClick={() => setActiveItemAndCloseSidePanel(NavbarItems.login)}
        >
          Login
        </Menu.Item>
      </Link>
    </Sidebar>
  );

  return (
    <>
      {isMobileBreakpoint ? (
        <>
          {mobileTopMenuItems}
          {mobileSideMenuItems}
        </>
      ) : (
        desktopMenuItems
      )}
    </>
  );
};

export default NavigationBar;
