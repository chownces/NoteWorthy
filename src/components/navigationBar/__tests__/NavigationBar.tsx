import { render } from '@testing-library/react';
import { Context as ResponsiveContext } from 'react-responsive';
import { MemoryRouter } from 'react-router-dom';

import NavigationBar from '../NavigationBar';

test('Navigation Bar renders correctly at desktop breakpoint', () => {
  const { container: desktop } = render(
    <ResponsiveContext.Provider value={{ width: 1000 }}>
      <MemoryRouter>
        <NavigationBar />
      </MemoryRouter>
    </ResponsiveContext.Provider>
  );
  expect(desktop).toMatchSnapshot();
});

test('Navigation Bar renders correctly at mobile breakpoint', () => {
  const { container: mobile } = render(
    <ResponsiveContext.Provider value={{ width: 600 }}>
      <MemoryRouter>
        <NavigationBar />
      </MemoryRouter>
    </ResponsiveContext.Provider>
  );
  expect(mobile).toMatchSnapshot();
});
