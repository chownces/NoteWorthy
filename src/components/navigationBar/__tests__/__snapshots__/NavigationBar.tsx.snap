// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Bar renders correctly at desktop breakpoint 1`] = `
<div>
  <div
    class="ui desktop-navbar menu"
  >
    <a
      class="brand-logo-link"
      href="/"
    >
      <div
        class="item brand-logo-container"
      >
        <img
          class="ui small image"
          src="brand_logo.png"
        />
      </div>
    </a>
    <a
      aria-current="page"
      class="active"
      href="/"
    >
      <div
        class="item"
      >
        <i
          aria-hidden="true"
          class="file alternate outline icon"
        />
        All Databases
      </div>
    </a>
    <a
      href="/todos"
    >
      <div
        class="item"
      >
        <i
          aria-hidden="true"
          class="calendar check outline icon"
        />
        Todo List
      </div>
    </a>
    <div
      class="right menu"
    >
      <a
        href="/contribute"
      >
        <div
          class="item"
        >
          Contribute
        </div>
      </a>
      <a
        href="/login"
      >
        <div
          class="item"
        >
          Login
        </div>
      </a>
    </div>
  </div>
</div>
`;

exports[`Navigation Bar renders correctly at mobile breakpoint 1`] = `
<div>
  <div
    class="ui menu"
  >
    <a
      class="item"
    >
      <i
        aria-hidden="true"
        class="bars icon"
      />
    </a>
  </div>
  <div
    class="ui vertical ui overlay left sidebar mobile-navbar menu"
  >
    <a
      href="/database/root"
    >
      <div
        class="item"
      >
        <i
          aria-hidden="true"
          class="file alternate outline icon"
        />
        All Databases
      </div>
    </a>
    <a
      href="/todos"
    >
      <div
        class="item"
      >
        <i
          aria-hidden="true"
          class="calendar check outline icon"
        />
        Todo List
      </div>
    </a>
    <a
      href="/contribute"
    >
      <div
        class="item"
      >
        Contribute
      </div>
    </a>
    <a
      href="/login"
    >
      <div
        class="item"
      >
        Login
      </div>
    </a>
  </div>
</div>
`;
